File: experiments/leopard2/code_identity/code_identity.c
/* Independent C99 oracle for the experimental Leopard2 code identity. */
#include "code_identity.h"

#include <limits.h>
#include <string.h>

static const uint8_t k_magic[4] = { 'L', '2', 'I', 'D' };

static uint16_t load_be16(const uint8_t *input)
{
    return (uint16_t)(((uint16_t)input[0] << 8) | (uint16_t)input[1]);
}

static uint32_t load_be32(const uint8_t *input)
{
    return ((uint32_t)input[0] << 24) |
        ((uint32_t)input[1] << 16) |
        ((uint32_t)input[2] << 8) |
        (uint32_t)input[3];
}

static void store_be16(uint8_t *output, uint16_t value)
{
    output[0] = (uint8_t)(value >> 8);
    output[1] = (uint8_t)value;
}

static void store_be32(uint8_t *output, uint32_t value)
{
    output[0] = (uint8_t)(value >> 24);
    output[1] = (uint8_t)(value >> 16);
    output[2] = (uint8_t)(value >> 8);
    output[3] = (uint8_t)value;
}

static int known_critical(uint16_t type)
{
    return type >= LEO2_CODE_ID_META_PROFILE_PARAMETERS_SHA256 &&
        type <= LEO2_CODE_ID_META_SHARD_LAYOUT;
}

static int known_digest(uint16_t type)
{
    return type >= LEO2_CODE_ID_META_PROFILE_PARAMETERS_SHA256 &&
        type <= LEO2_CODE_ID_META_PUNCTURING_SET_SHA256;
}

static int aliases_known_critical(uint16_t type)
{
    return (type & UINT16_C(0x8000)) == 0u &&
        known_critical(type | UINT16_C(0x8000));
}

static int ceil_pow2_bounded(uint64_t value, uint32_t *result)
{
    uint64_t power = 1;
    if (value == 0 || result == NULL) {
        return 0;
    }
    while (power < value) {
        if (power > UINT64_MAX / 2u) {
            return 0;
        }
        power *= 2u;
    }
    if (power > UINT32_MAX) {
        return 0;
    }
    *result = (uint32_t)power;
    return 1;
}

static leo2_code_id_status derive_counts(
    uint8_t profile,
    uint32_t original_count,
    uint32_t recovery_count,
    uint32_t *parent_count,
    uint32_t *padded_side)
{
    uint64_t parent_input;
    uint32_t side;

    if (original_count == 0 || recovery_count == 0 ||
        parent_count == NULL || padded_side == NULL) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }
    if (profile == LEO2_CODE_ID_PROFILE_LEGACY_HIGH) {
        if (!ceil_pow2_bounded(recovery_count, &side)) {
            return LEO2_CODE_ID_INVALID_IDENTITY;
        }
        parent_input = (uint64_t)original_count + (uint64_t)side;
    } else if (profile == LEO2_CODE_ID_PROFILE_LOW) {
        if (!ceil_pow2_bounded(original_count, &side)) {
            return LEO2_CODE_ID_INVALID_IDENTITY;
        }
        parent_input = (uint64_t)side + (uint64_t)recovery_count;
    } else if (profile == LEO2_CODE_ID_PROFILE_EXACT_LOW) {
        parent_input = (uint64_t)original_count + (uint64_t)recovery_count;
        if (parent_input > UINT32_MAX) {
            return LEO2_CODE_ID_INVALID_IDENTITY;
        }
        *parent_count = (uint32_t)parent_input;
        *padded_side = original_count;
        return LEO2_CODE_ID_OK;
    } else {
        return LEO2_CODE_ID_UNSUPPORTED;
    }
    if (!ceil_pow2_bounded(parent_input, parent_count)) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }
    *padded_side = side;
    return LEO2_CODE_ID_OK;
}

leo2_code_id_status leo2_code_identity_validate(
    const leo2_code_identity *identity)
{
    uint32_t derived_parent;
    uint32_t derived_side;
    uint16_t previous = 0;
    size_t index;
    leo2_code_id_status status;

    if (identity == NULL) {
        return LEO2_CODE_ID_INVALID_ARGUMENT;
    }
    if (identity->profile != LEO2_CODE_ID_PROFILE_LEGACY_HIGH &&
        identity->profile != LEO2_CODE_ID_PROFILE_LOW &&
        identity->profile != LEO2_CODE_ID_PROFILE_EXACT_LOW) {
        return LEO2_CODE_ID_UNSUPPORTED;
    }
    if (identity->profile_version != 1u) {
        return LEO2_CODE_ID_UNSUPPORTED;
    }
    if (identity->field != LEO2_CODE_ID_FIELD_GF8 &&
        identity->field != LEO2_CODE_ID_FIELD_GF16) {
        return LEO2_CODE_ID_UNSUPPORTED;
    }
    if (identity->field_version != 1u ||
        identity->coordinate_map_version != 1u) {
        return LEO2_CODE_ID_UNSUPPORTED;
    }

    status = derive_counts(identity->profile, identity->original_count,
        identity->recovery_count, &derived_parent, &derived_side);
    if (status != LEO2_CODE_ID_OK) {
        return status;
    }
    if (derived_parent != identity->parent_count ||
        derived_side != identity->padded_side) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }
    if (derived_parent > 65536u) {
        return LEO2_CODE_ID_UNSUPPORTED;
    }
    if (identity->field == LEO2_CODE_ID_FIELD_GF8 && derived_parent > 256u) {
        return LEO2_CODE_ID_UNSUPPORTED;
    }
    if (identity->metadata_count > LEO2_CODE_ID_MAX_METADATA) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }

    for (index = 0; index < identity->metadata_count; ++index) {
        const leo2_code_id_metadata *item = &identity->metadata[index];
        size_t prior;
        if (item->type == 0u) {
            return LEO2_CODE_ID_NONCANONICAL;
        }
        if (index != 0u && item->type <= previous) {
            return LEO2_CODE_ID_NONCANONICAL;
        }
        previous = item->type;
        for (prior = 0; prior < index; ++prior) {
            if ((identity->metadata[prior].type & UINT16_C(0x7fff)) ==
                (item->type & UINT16_C(0x7fff))) {
                return LEO2_CODE_ID_NONCANONICAL;
            }
        }
        if (aliases_known_critical(item->type)) {
            return LEO2_CODE_ID_NONCANONICAL;
        }
        if ((item->type & UINT16_C(0x8000)) != 0u &&
            !known_critical(item->type)) {
            return LEO2_CODE_ID_UNSUPPORTED;
        }
        if (item->value_bytes > LEO2_CODE_ID_MAX_METADATA_VALUE) {
            return LEO2_CODE_ID_INVALID_IDENTITY;
        }
        if (item->value_bytes != 0u && item->value == NULL) {
            return LEO2_CODE_ID_INVALID_ARGUMENT;
        }
        if (known_digest(item->type) && item->value_bytes != 32u) {
            return LEO2_CODE_ID_INVALID_IDENTITY;
        }
        if (identity->profile == LEO2_CODE_ID_PROFILE_EXACT_LOW &&
            (item->type == LEO2_CODE_ID_META_SHORTENING_SET_SHA256 ||
             item->type == LEO2_CODE_ID_META_PUNCTURING_SET_SHA256)) {
            return LEO2_CODE_ID_INVALID_IDENTITY;
        }
        if (item->type == LEO2_CODE_ID_META_SHARD_LAYOUT) {
            uint8_t layout;
            if (item->value_bytes != 1u || item->value == NULL) {
                return LEO2_CODE_ID_INVALID_IDENTITY;
            }
            layout = item->value[0];
            if (layout == LEO2_CODE_ID_SHARD_LAYOUT_NATIVE_V1) {
                return LEO2_CODE_ID_NONCANONICAL;
            }
            if (layout != LEO2_CODE_ID_SHARD_LAYOUT_GF16_PADDED_ODD_V1) {
                return LEO2_CODE_ID_UNSUPPORTED;
            }
            if (identity->field != LEO2_CODE_ID_FIELD_GF16) {
                return LEO2_CODE_ID_INVALID_IDENTITY;
            }
        }
    }
    return LEO2_CODE_ID_OK;
}

leo2_code_id_status leo2_code_identity_make(
    leo2_code_identity *identity,
    uint8_t profile,
    uint8_t field,
    uint32_t original_count,
    uint32_t recovery_count)
{
    leo2_code_id_status status;
    if (identity == NULL) {
        return LEO2_CODE_ID_INVALID_ARGUMENT;
    }
    memset(identity, 0, sizeof(*identity));
    identity->profile = profile;
    identity->profile_version = 1u;
    identity->field = field;
    identity->field_version = 1u;
    identity->original_count = original_count;
    identity->recovery_count = recovery_count;
    identity->coordinate_map_version = 1u;
    status = derive_counts(profile, original_count, recovery_count,
        &identity->parent_count, &identity->padded_side);
    if (status != LEO2_CODE_ID_OK) {
        return status;
    }
    return leo2_code_identity_validate(identity);
}

leo2_code_id_status leo2_code_identity_add_metadata(
    leo2_code_identity *identity,
    uint16_t type,
    const uint8_t *value,
    size_t value_bytes)
{
    size_t position;
    size_t index;
    leo2_code_id_metadata item;

    if (identity == NULL || (value_bytes != 0u && value == NULL)) {
        return LEO2_CODE_ID_INVALID_ARGUMENT;
    }
    if (type == 0u || value_bytes > LEO2_CODE_ID_MAX_METADATA_VALUE ||
        value_bytes > UINT16_MAX) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }
    if (identity->metadata_count >= LEO2_CODE_ID_MAX_METADATA) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }
    if (aliases_known_critical(type)) {
        return LEO2_CODE_ID_NONCANONICAL;
    }
    if ((type & UINT16_C(0x8000)) != 0u && !known_critical(type)) {
        return LEO2_CODE_ID_UNSUPPORTED;
    }
    if (known_digest(type) && value_bytes != 32u) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }
    if (identity->profile == LEO2_CODE_ID_PROFILE_EXACT_LOW &&
        (type == LEO2_CODE_ID_META_SHORTENING_SET_SHA256 ||
         type == LEO2_CODE_ID_META_PUNCTURING_SET_SHA256)) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }
    if (type == LEO2_CODE_ID_META_SHARD_LAYOUT) {
        uint8_t layout;
        if (value_bytes != 1u) {
            return LEO2_CODE_ID_INVALID_IDENTITY;
        }
        layout = value[0];
        if (layout == LEO2_CODE_ID_SHARD_LAYOUT_NATIVE_V1) {
            return LEO2_CODE_ID_NONCANONICAL;
        }
        if (layout != LEO2_CODE_ID_SHARD_LAYOUT_GF16_PADDED_ODD_V1) {
            return LEO2_CODE_ID_UNSUPPORTED;
        }
        if (identity->field != LEO2_CODE_ID_FIELD_GF16) {
            return LEO2_CODE_ID_INVALID_IDENTITY;
        }
    }
    for (index = 0; index < identity->metadata_count; ++index) {
        if ((identity->metadata[index].type & UINT16_C(0x7fff)) ==
            (type & UINT16_C(0x7fff))) {
            return LEO2_CODE_ID_NONCANONICAL;
        }
    }

    position = 0;
    while (position < identity->metadata_count &&
        identity->metadata[position].type < type) {
        ++position;
    }
    for (index = identity->metadata_count; index > position; --index) {
        identity->metadata[index] = identity->metadata[index - 1u];
    }
    item.type = type;
    item.value_bytes = (uint16_t)value_bytes;
    item.value = value;
    identity->metadata[position] = item;
    ++identity->metadata_count;
    return leo2_code_identity_validate(identity);
}

leo2_code_id_status leo2_code_identity_serialize(
    const leo2_code_identity *identity,
    uint8_t *output,
    size_t output_capacity,
    size_t *required_bytes)
{
    size_t total = LEO2_CODE_ID_HEADER_BYTES;
    size_t offset;
    size_t index;
    leo2_code_id_status status;

    if (required_bytes == NULL) {
        return LEO2_CODE_ID_INVALID_ARGUMENT;
    }
    *required_bytes = 0;
    status = leo2_code_identity_validate(identity);
    if (status != LEO2_CODE_ID_OK) {
        return status;
    }
    for (index = 0; index < identity->metadata_count; ++index) {
        total += 4u + identity->metadata[index].value_bytes;
    }
    if (total > LEO2_CODE_ID_MAX_BYTES) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }
    *required_bytes = total;
    if (output == NULL) {
        return output_capacity == 0u ? LEO2_CODE_ID_OK :
            LEO2_CODE_ID_INVALID_ARGUMENT;
    }
    if (output_capacity < total) {
        return LEO2_CODE_ID_BUFFER_TOO_SMALL;
    }

    memcpy(output, k_magic, sizeof(k_magic));
    output[4] = 1u;
    output[5] = 0u;
    store_be16(output + 6, LEO2_CODE_ID_HEADER_BYTES);
    store_be32(output + 8, (uint32_t)total);
    output[12] = identity->profile;
    output[13] = identity->profile_version;
    output[14] = identity->field;
    output[15] = identity->field_version;
    store_be32(output + 16, identity->original_count);
    store_be32(output + 20, identity->recovery_count);
    store_be32(output + 24, identity->parent_count);
    store_be32(output + 28, identity->padded_side);
    store_be16(output + 32, identity->coordinate_map_version);
    store_be16(output + 34, identity->metadata_count);

    offset = LEO2_CODE_ID_HEADER_BYTES;
    for (index = 0; index < identity->metadata_count; ++index) {
        const leo2_code_id_metadata *item = &identity->metadata[index];
        store_be16(output + offset, item->type);
        store_be16(output + offset + 2u, item->value_bytes);
        offset += 4u;
        if (item->value_bytes != 0u) {
            memcpy(output + offset, item->value, item->value_bytes);
            offset += item->value_bytes;
        }
    }
    return LEO2_CODE_ID_OK;
}

leo2_code_id_status leo2_code_identity_deserialize(
    const uint8_t *input,
    size_t input_bytes,
    leo2_code_identity *identity)
{
    uint32_t total_bytes;
    uint16_t metadata_count;
    size_t offset;
    size_t index;
    leo2_code_id_status status;

    if (input == NULL || identity == NULL) {
        return LEO2_CODE_ID_INVALID_ARGUMENT;
    }
    memset(identity, 0, sizeof(*identity));
    if (input_bytes < LEO2_CODE_ID_HEADER_BYTES) {
        return LEO2_CODE_ID_TRUNCATED;
    }
    if (memcmp(input, k_magic, sizeof(k_magic)) != 0) {
        return LEO2_CODE_ID_NONCANONICAL;
    }
    if (input[4] != 1u) {
        return LEO2_CODE_ID_UNSUPPORTED;
    }
    if (input[5] != 0u || load_be16(input + 6) != LEO2_CODE_ID_HEADER_BYTES) {
        return LEO2_CODE_ID_NONCANONICAL;
    }
    total_bytes = load_be32(input + 8);
    if ((uint64_t)total_bytes != (uint64_t)input_bytes ||
        total_bytes > LEO2_CODE_ID_MAX_BYTES) {
        return LEO2_CODE_ID_NONCANONICAL;
    }
    metadata_count = load_be16(input + 34);
    if (metadata_count > LEO2_CODE_ID_MAX_METADATA) {
        return LEO2_CODE_ID_INVALID_IDENTITY;
    }

    identity->profile = input[12];
    identity->profile_version = input[13];
    identity->field = input[14];
    identity->field_version = input[15];
    identity->original_count = load_be32(input + 16);
    identity->recovery_count = load_be32(input + 20);
    identity->parent_count = load_be32(input + 24);
    identity->padded_side = load_be32(input + 28);
    identity->coordinate_map_version = load_be16(input + 32);
    identity->metadata_count = metadata_count;

    offset = LEO2_CODE_ID_HEADER_BYTES;
    for (index = 0; index < metadata_count; ++index) {
        uint16_t value_bytes;
        if (input_bytes - offset < 4u) {
            return LEO2_CODE_ID_TRUNCATED;
        }
        identity->metadata[index].type = load_be16(input + offset);
        value_bytes = load_be16(input + offset + 2u);
        identity->metadata[index].value_bytes = value_bytes;
        offset += 4u;
        if (value_bytes > LEO2_CODE_ID_MAX_METADATA_VALUE ||
            (size_t)value_bytes > input_bytes - offset) {
            return LEO2_CODE_ID_TRUNCATED;
        }
        identity->metadata[index].value = input + offset;
        offset += value_bytes;
    }
    if (offset != input_bytes) {
        return LEO2_CODE_ID_NONCANONICAL;
    }
    status = leo2_code_identity_validate(identity);
    if (status != LEO2_CODE_ID_OK) {
        return status;
    }
    return LEO2_CODE_ID_OK;
}

const char *leo2_code_id_status_string(leo2_code_id_status status)
{
    switch (status) {
    case LEO2_CODE_ID_OK: return "ok";
    case LEO2_CODE_ID_INVALID_ARGUMENT: return "invalid argument";
    case LEO2_CODE_ID_INVALID_IDENTITY: return "invalid identity";
    case LEO2_CODE_ID_UNSUPPORTED: return "unsupported semantic extension";
    case LEO2_CODE_ID_NONCANONICAL: return "noncanonical input";
    case LEO2_CODE_ID_TRUNCATED: return "truncated input";
    case LEO2_CODE_ID_BUFFER_TOO_SMALL: return "buffer too small";
    default: return "unknown status";
    }
}

File: experiments/leopard2/code_identity/code_identity.py
#!/usr/bin/env python3
"""Experimental Leopard2 wire-code identity serializer.

This is deliberately not part of the public API.  It is a small independent
reference used to freeze and test the Experiment W binary proposal.
"""

from __future__ import annotations

import dataclasses
import struct
from typing import Iterable, Tuple


MAGIC = b"L2ID"
FORMAT_VERSION = 1
HEADER_BYTES = 36
MAX_ID_BYTES = 65535
MAX_METADATA = 64
MAX_METADATA_VALUE = 4096

PROFILE_LEGACY_HIGH = 1
PROFILE_LOW = 2
# This is the persistent identity assigned to the public enum value
# LEO2_PROFILE_EXACT_EXPERIMENTAL_V1.  C7 freezes its v1 coordinate semantics
# as exact-prefix low-rate; production codec construction remains unsupported.
PROFILE_EXACT_LOW = 3
# Family 4 is deliberately not accepted yet.  Reserving the number here keeps
# C8's future exact-high profile from reusing family 3 or silently changing the
# public enum-3 exact-low mathematics.
PROFILE_EXACT_HIGH_RESERVED = 4
FIELD_GF8 = 1
FIELD_GF16 = 2

# Critical metadata known by this reader.  The first four values are SHA-256
# digests of canonical mathematical objects; shard layout is a one-byte wire
# framing selector.  None of these values names an implementation choice.
META_PROFILE_PARAMETERS_SHA256 = 0x8001
META_COORDINATE_SET_SHA256 = 0x8002
META_SHORTENING_SET_SHA256 = 0x8003
META_PUNCTURING_SET_SHA256 = 0x8004
META_SHARD_LAYOUT = 0x8005
SHARD_LAYOUT_NATIVE_V1 = 0
SHARD_LAYOUT_GF16_PADDED_ODD_V1 = 1
_KNOWN_DIGESTS = {
    META_PROFILE_PARAMETERS_SHA256,
    META_COORDINATE_SET_SHA256,
    META_SHORTENING_SET_SHA256,
    META_PUNCTURING_SET_SHA256,
}
_KNOWN_CRITICAL = _KNOWN_DIGESTS | {META_SHARD_LAYOUT}


class IdentityError(ValueError):
    """The bytes are malformed, non-canonical, or semantically unsupported."""


@dataclasses.dataclass(frozen=True, order=True)
class Metadata:
    type: int
    value: bytes


@dataclasses.dataclass(frozen=True)
class CodeIdentity:
    profile: int
    profile_version: int
    field: int
    field_version: int
    original_count: int
    recovery_count: int
    parent_count: int
    padded_side: int
    coordinate_map_version: int
    metadata: Tuple[Metadata, ...] = ()


_HEADER = struct.Struct(">4sBBHI4B4I2H")
_TLV = struct.Struct(">HH")
if _HEADER.size != HEADER_BYTES:
    raise RuntimeError("code-identity header size constant is inconsistent")


def _ceil_pow2(value: int) -> int:
    if value <= 0:
        raise IdentityError("counts must be positive")
    return 1 << (value - 1).bit_length()


def _validate(identity: CodeIdentity) -> None:
    scalar_u8 = (
        ("profile", identity.profile),
        ("profile_version", identity.profile_version),
        ("field", identity.field),
        ("field_version", identity.field_version),
    )
    for name, value in scalar_u8:
        if not isinstance(value, int) or not 0 <= value <= 0xFF:
            raise IdentityError(f"{name} does not fit uint8")
    scalar_u32 = (
        ("original_count", identity.original_count),
        ("recovery_count", identity.recovery_count),
        ("parent_count", identity.parent_count),
        ("padded_side", identity.padded_side),
    )
    for name, value in scalar_u32:
        if not isinstance(value, int) or not 0 <= value <= 0xFFFFFFFF:
            raise IdentityError(f"{name} does not fit uint32")
    if not isinstance(identity.coordinate_map_version, int) or not (
        0 <= identity.coordinate_map_version <= 0xFFFF
    ):
        raise IdentityError("coordinate_map_version does not fit uint16")

    if identity.profile not in (
        PROFILE_LEGACY_HIGH, PROFILE_LOW, PROFILE_EXACT_LOW
    ):
        raise IdentityError("unknown profile family")
    if identity.profile_version != 1:
        raise IdentityError("unsupported profile version")
    if identity.field not in (FIELD_GF8, FIELD_GF16):
        raise IdentityError("unknown field")
    if identity.field_version != 1:
        raise IdentityError("unsupported field representation version")
    if identity.coordinate_map_version != 1:
        raise IdentityError("unsupported coordinate-map version")
    if identity.original_count == 0 or identity.recovery_count == 0:
        raise IdentityError("K and R must be positive")

    if identity.profile == PROFILE_LEGACY_HIGH:
        side = _ceil_pow2(identity.recovery_count)
        parent = _ceil_pow2(identity.original_count + side)
    elif identity.profile == PROFILE_LOW:
        side = _ceil_pow2(identity.original_count)
        parent = _ceil_pow2(side + identity.recovery_count)
    else:
        # Exact-low V1 has no dyadic parent or padded side.  These redundant
        # envelope fields carry its exact transmitted length and K-side size.
        side = identity.original_count
        parent = identity.original_count + identity.recovery_count
        if parent > 0xFFFFFFFF:
            raise IdentityError("exact code length overflows uint32")
    if identity.padded_side != side or identity.parent_count != parent:
        raise IdentityError("parent or padded side is inconsistent with profile")
    if parent > 65536:
        raise IdentityError("parent exceeds supported field order")
    if identity.field == FIELD_GF8 and parent > 256:
        raise IdentityError("GF8 parent exceeds 256 coordinates")

    if len(identity.metadata) > MAX_METADATA:
        raise IdentityError("too many metadata entries")
    previous = -1
    seen_base_types = set()
    for item in identity.metadata:
        if not isinstance(item, Metadata):
            raise IdentityError("metadata entries must be Metadata")
        if not 0 < item.type <= 0xFFFF:
            raise IdentityError("metadata type zero is reserved")
        if item.type <= previous:
            raise IdentityError("metadata is not in canonical type order")
        previous = item.type
        base_type = item.type & 0x7FFF
        if base_type in seen_base_types:
            raise IdentityError("duplicate metadata base type")
        seen_base_types.add(base_type)
        if not item.type & 0x8000 and (item.type | 0x8000) in _KNOWN_CRITICAL:
            raise IdentityError("noncritical alias of known critical metadata")
        if item.type & 0x8000 and item.type not in _KNOWN_CRITICAL:
            raise IdentityError("unknown critical metadata")
        if not isinstance(item.value, bytes):
            raise IdentityError("metadata values must be bytes")
        if len(item.value) > MAX_METADATA_VALUE:
            raise IdentityError("metadata value is too long")
        if item.type in _KNOWN_DIGESTS and len(item.value) != 32:
            raise IdentityError("known digest metadata must be 32 bytes")
        if identity.profile == PROFILE_EXACT_LOW and item.type in (
            META_COORDINATE_SET_SHA256,
            META_SHORTENING_SET_SHA256,
            META_PUNCTURING_SET_SHA256,
        ):
            raise IdentityError(
                "exact-low V1 map is fixed and has no coordinate/shortening/puncturing TLV"
            )
        if item.type == META_SHARD_LAYOUT:
            if len(item.value) != 1:
                raise IdentityError("shard-layout metadata must be one byte")
            layout = item.value[0]
            if layout == SHARD_LAYOUT_NATIVE_V1:
                raise IdentityError("explicit native shard layout is non-canonical")
            if layout != SHARD_LAYOUT_GF16_PADDED_ODD_V1:
                raise IdentityError("unsupported shard layout")
            if identity.field != FIELD_GF16:
                raise IdentityError("padded-odd shard layout requires GF16")


def serialize(identity: CodeIdentity) -> bytes:
    """Return the one canonical v1 representation of identity."""
    _validate(identity)
    total = HEADER_BYTES + sum(_TLV.size + len(item.value) for item in identity.metadata)
    if total > MAX_ID_BYTES:
        raise IdentityError("serialized identity is too long")
    result = bytearray(
        _HEADER.pack(
            MAGIC,
            FORMAT_VERSION,
            0,  # reserved flags: must stay zero in v1
            HEADER_BYTES,
            total,
            identity.profile,
            identity.profile_version,
            identity.field,
            identity.field_version,
            identity.original_count,
            identity.recovery_count,
            identity.parent_count,
            identity.padded_side,
            identity.coordinate_map_version,
            len(identity.metadata),
        )
    )
    for item in identity.metadata:
        result += _TLV.pack(item.type, len(item.value))
        result += item.value
    return bytes(result)


def deserialize(data: bytes) -> CodeIdentity:
    """Parse a canonical v1 identity, failing closed on semantic extensions."""
    if not isinstance(data, bytes):
        raise IdentityError("serialized identity must be bytes")
    if len(data) < HEADER_BYTES:
        raise IdentityError("truncated header")
    (
        magic,
        format_version,
        flags,
        header_bytes,
        total_bytes,
        profile,
        profile_version,
        field,
        field_version,
        original_count,
        recovery_count,
        parent_count,
        padded_side,
        coordinate_map_version,
        metadata_count,
    ) = _HEADER.unpack_from(data)
    if magic != MAGIC:
        raise IdentityError("bad magic")
    if format_version != FORMAT_VERSION:
        raise IdentityError("unsupported envelope version")
    if flags != 0:
        raise IdentityError("reserved flags are nonzero")
    if header_bytes != HEADER_BYTES:
        raise IdentityError("non-canonical header size")
    if total_bytes != len(data) or total_bytes > MAX_ID_BYTES:
        raise IdentityError("declared length does not match input")
    if metadata_count > MAX_METADATA:
        raise IdentityError("too many metadata entries")

    metadata = []
    offset = HEADER_BYTES
    for _ in range(metadata_count):
        if len(data) - offset < _TLV.size:
            raise IdentityError("truncated metadata header")
        type_id, length = _TLV.unpack_from(data, offset)
        offset += _TLV.size
        if length > MAX_METADATA_VALUE or length > len(data) - offset:
            raise IdentityError("truncated or oversized metadata value")
        metadata.append(Metadata(type_id, data[offset : offset + length]))
        offset += length
    if offset != len(data):
        raise IdentityError("unaccounted trailing bytes")

    identity = CodeIdentity(
        profile,
        profile_version,
        field,
        field_version,
        original_count,
        recovery_count,
        parent_count,
        padded_side,
        coordinate_map_version,
        tuple(metadata),
    )
    _validate(identity)
    if serialize(identity) != data:
        raise IdentityError("non-canonical representation")
    return identity


def make_identity(
    profile: int,
    field: int,
    original_count: int,
    recovery_count: int,
    metadata: Iterable[Metadata] = (),
) -> CodeIdentity:
    """Construct a current-profile identity and derive its redundant fields."""
    if profile == PROFILE_LEGACY_HIGH:
        side = _ceil_pow2(recovery_count)
        parent = _ceil_pow2(original_count + side)
    elif profile == PROFILE_LOW:
        side = _ceil_pow2(original_count)
        parent = _ceil_pow2(side + recovery_count)
    elif profile == PROFILE_EXACT_LOW:
        side = original_count
        parent = original_count + recovery_count
    else:
        raise IdentityError("unknown profile family")
    return CodeIdentity(
        profile=profile,
        profile_version=1,
        field=field,
        field_version=1,
        original_count=original_count,
        recovery_count=recovery_count,
        parent_count=parent,
        padded_side=side,
        coordinate_map_version=1,
        metadata=tuple(sorted(metadata)),
    )
